File: Python/Algorithms/CountingSort.py


class CountingSort:

    def __init__(self, data):
        self.data = data
        self.count_array = [0 for _ in range(max(data) - min(data) + 1)]

    def sort(self):
        min_value = min(self.data)
        for i in range(len(self.data)):
            self.count_array[self.data[i] - min_value] += 1

        z = 0
        for i in range(min(self.data), max(self.data) + 1):
            while self.count_array[i - min_value] > 0:
                self.data[z] = i
                z += 1
                self.count_array[i-min_value] -= 1


if __name__ == '__main__':
    n = [4, 6, 3, 43, 1, -4]
    counting_sort = CountingSort(n)
    counting_sort.sort()
    print(counting_sort.data)

File: C#/Algorithms/Interview/CountingSort.cs
using System;
using System.Linq;

namespace Algorithms.Interview
{
    public class CountingSort
    {
        private int[] nums;
        int[] countArray;
        int minValue;
        int maxValue;
        public CountingSort(int[] nums)
        {
            this.nums = nums;
            minValue = nums.Min();
            maxValue = nums.Max();
            countArray = new int[maxValue - minValue + 1];
        }

        public void Sort()
        {
            for(int i = 0; i < nums.Length; i++)
            {
                countArray[nums[i] - minValue] += 1;
            }

            int z = 0;
            for(int i = minValue; i < maxValue + 1; i++)
            {
                while(countArray[i - minValue] > 0)
                {
                    nums[z] = i;
                    z++;
                    countArray[i - minValue] -= 1;
                }
            }
        }

        public void PrintNums()
        {
            for (int i = 0; i < nums.Length; i++)
            {
                Console.Write($"{nums[i]},");
            }
            Console.WriteLine();
        }
    }
}
